manticore/core/executor.py

```diff
@@ -69,184 +46,271 @@ def newFunction(self, *args, **kw):
+        self._test_count.value += 1
+        return self._test_count.value
 
-    @dump_every.setter
-    def dump_every(self, val):
-        assert(isinstance(val, (int, long)))
-        self._dump_every = val
+    ###############################################
+    # Synchronization helpers
+    @sync
+    def _start_run(self):
+        #notify siblings we are about to start a run()
+        self._running.value+=1
 
     @sync
-    def delState(self, state):
-        return self._delState(state)
+    def _stop_run(self, count=0):
+        #notify siblings we are about to stop this run()
+        self._running.value-=1
+        assert self._running.value >=0
+        self._lock.notify_all()
+
+
+    ################################################
+    #Public API
+    @property
+    def running(self):
+        ''' Report an estimate  of how many workers are currently running '''
+        return self._running.value
+
+    def shutdown(self):
+        ''' This will stop all workers '''
+        self._shutdown.set()
+
+    def is_shutdown(self):
+        ''' Returns True if shutdown was requested '''
+        return self._shutdown.is_set()
 
-    def _delState(self, state):
-        logger.debug("Removing state %d from storage", state.co)
-        assert state.name in self._states
-        del self._states[state.name]
-        os.remove(self._getFilename(state.name))
 
+    ###############################################
+    # Priority queue 
     @sync
-    def putState(self, state):
-        '''
-        Serialize and store a given state.
+    def put(self, state_id):
+        ''' Enqueue it for processing '''
+        self._states.append(state_id)
+        self._lock.notify_all()
+        return state_id
 
-        :param state: The state to serialize
-        '''
-        return self._putState(state)
+    @sync
+    def get(self):
+        ''' Dequeue a state with the max priority '''
 
-    def _putState(self, state):
-        logger.debug("Saving state %d to file %s", state.co, state.name)
+        #A shutdown has been requested
+        if self.is_shutdown():
+            return None
 
-        with open(self._getFilename(state.name), 'w+') as f:
+        #if not more states in the queue lets wait for some forks
+        while len(self._states) == 0:
+            #if no worker is running bail out
+            if self.running == 0:
+                return None
+            #if a shutdown has been requested bail out
+            if self.is_shutdown():
+                return None
+            #if there is actually some workers running wait for state forks
+            logger.debug("Waiting for available states")
+            self._lock.wait()
+            
+        state_id = random.choice(self._states)
+        del  self._states[self._states.index(state_id)]
+        return state_id
+
+    ###############################################################
+    # File Storage 
+    def backup(self, state):
```

Comment:
Its when you save a State to the secondary storage(in our current implementation is a pickled file on the workspace folder).    backup/restore